core/Pos/Core/Slotting/Timestamp.hs

```diff
@@ -50,6 +60,28 @@ instance NFData Timestamp where
 timestampF :: Format r (Timestamp -> r)
 timestampF = build
 
+-- | Attempt to parse a 'Timestamp' out of a 'Text' value. Formats include:
+--
+-- * Fractional timestamps: @123456789.1234@
+-- * ISO8601 Datetime: @1999-10-12T08:09:10@
+-- * ISO8601 Date: @1999-10-12@
+parseTimestamp :: Text -> Maybe Timestamp
+parseTimestamp t = utcTimeParser <|> timePosixParser
+  where
+    str = Text.unpack t
+    parseFmt :: String -> Maybe UTCTime
+    parseFmt fmt =
+        parseTimeM True defaultTimeLocale fmt str
+    utcTimeParser =
+        view (from timestampToUTCTimeL)
+        <$> asum
+            [ parseFmt (iso8601DateFormat (Just "%H:%M:%S%Q"))
+            , parseFmt (iso8601DateFormat Nothing)
+            ]
+    timePosixParser =
+        view (from timestampSeconds)
+        <$> readMaybe @Double str
```

Comment:
I have factored this to a more central location for use in both the `FromJSON` instance and the `ToIndex` instance.